scripts/install-cni.sh

```diff
@@ -130,20 +130,149 @@ else
+    fi
+  done
+
+  # Directpath use case
+  if [ "$ENABLE_IPV6" == "true" ] && [ "$ipv6_subnet_configured" == "false" ] ; then
+    # Directpath adds IPv6 subnet and route derived from host with fixed range
+    # of /112 even when it is not specified in node's .spec.podCIDRs
+    if [ -n "${node_ipv6_addr:-}" ] && [ "${node_ipv6_addr}" != "null" ]; then
+      POPULATE_IP6TABLES="true"
+      SUBNETS_REPLACEMENT+=("[{\"subnet\": \"${node_ipv6_addr}/112\"}]")
+      ROUTES_REPLACEMENT+=("${CNI_SPEC_IPV6_ROUTE:-{\"dst\": \"::/0\"\}}")
+    fi
+  fi
+
+  SUBNETS_REPLACEMENT_CONCATENATED=$(IFS=', ' ; echo "${SUBNETS_REPLACEMENT[*]}")
+  ROUTES_REPLACEMENT_CONCATENATED=$(IFS=', ' ; echo "${ROUTES_REPLACEMENT[*]}")
+
+  cni_spec=${cni_spec//@subnets/$SUBNETS_REPLACEMENT_CONCATENATED}
+  cni_spec=${cni_spec//@routes/$ROUTES_REPLACEMENT_CONCATENATED}
+}
+
+#######################################
+# Replaces `@ipv4Subnet', '@ipv6SubnetOptional` and `@ipv6RouteOptional` placeholders using `.spec.podCIDR` from node json and node's ipv6 from metadata server.
+# Arguments:
+#   node json from kube-apiserver
+#   node_ipv6_addr node's IPv6 address from GCE metadata server
+#######################################
+function fillSubnetsInCniSpecLegacyTemplate {
+  local node=$1
+  local node_ipv6_addr=$2
+
+  local primary_subnet
+  primary_subnet=$(jq -r '.spec.podCIDR' <<<"$response")
+
+  if is_ipv4_range "${primary_subnet:-}" ; then
+    echo "PodCIDR validation succeeded: ${primary_subnet:-}"
+    echo "Filling IPv4 subnet ${primary_subnet:-}"
+    cni_spec=${cni_spec//@ipv4Subnet/[{\"subnet\": \"${primary_subnet:-}\"\}]}
+  elif is_ipv6_range "${primary_subnet:-}" ; then
+    echo "Primary IPv6 pod range detected '${primary_subnet:-}'. It will only work with new spec template."
+  else
+    echo "Response from $node_url"
+    echo "$response"
+    echo "Failed to fetch/validate PodCIDR from K8s API server, primary_subnet=${primary_subnet:-}. Exiting (1)..."
+    exit 1
+  fi
+
+  if [ "$ENABLE_IPV6" == "true" ] || [ "${STACK_TYPE:-}" == "IPV4_IPV6" ]; then
+    if [ -n "${node_ipv6_addr:-}" ] && [ "${node_ipv6_addr}" != "null" ]; then
+      echo "Found nic0 IPv6 address ${node_ipv6_addr:-}. Filling IPv6 subnet and route..."
+      POPULATE_IP6TABLES="true"
+
+      cni_spec=${cni_spec//@ipv6SubnetOptional/, [{\"subnet\": \"${node_ipv6_addr:-}/112\"\}]}
+      cni_spec=${cni_spec//@ipv6RouteOptional/, ${CNI_SPEC_IPV6_ROUTE:-{\"dst\": \"::/0\"\}}}
+    else
+      echo "No IPv6 address found for nic0. Clearing IPv6 subnet and route..."
+      cni_spec=${cni_spec//@ipv6SubnetOptional/}
+      cni_spec=${cni_spec//@ipv6RouteOptional/}
+    fi
+  else
+    echo "Clearing IPv6 subnet and route given IPv6 access is disabled..."
+    cni_spec=${cni_spec//@ipv6SubnetOptional/}
+    cni_spec=${cni_spec//@ipv6RouteOptional/}
+  fi
+
+}
+
+function fillSubnetsInCniSpec {
+  fillSubnetsInCniSpecV2Template "$1" "$2"
+  fillSubnetsInCniSpecLegacyTemplate "$1" "$2"
```

Comment:
@jingyuanliang I ran this thought through @marqc , should we have template versions instead of relying on the parsers?

Reply:
I don't have strong opinion.

Reply:
I think my main concern is process the template twice, seems very brittle and easy to have subtle mistakes, if we can just add a sentinel that allows us to differentiate the template, we can add an args parameter to the template and have our internal versioning

https://github.com/containernetworking/cni/blob/spec-v0.4.0/SPEC.md#network-configuration

```
{
  "cniVersion": "0.4.0",
  "name": "dbnet",
  "type": "bridge",
  "bridge": "cni0",
  "args": {
    "labels" : {
        "netdVersion" : "1.0"
    }
  },
  "ipam": {
    "type": "host-local",
    "subnet": "10.1.0.0/16",
    "gateway": "10.1.0.1"
  },
  "dns": {
    "nameservers": [ "10.1.0.1" ]
  }
}
```

Reply:
Well when you said that I was thinking about a new env var. A new label doesn't work well because the template itself is not valid json and can't be easily parsed.

Reply:
env var sounds good

Reply:
Added handler for env `CNI_SPEC_TEMPLATE_VERSION` that fallbacks to legacy (current) template parser if not set.